torchbenchmark/score/compute_score.py

```diff
@@ -7,148 +7,151 @@
-    configurations the benchmark was run on.
-
-    For eg., if the benchmark was run in eval mode on a GPU in Torchscript JIT,
-                config = (train, cuda, jit)
-
-    This helper returns the score_db .
-
-    """
-    found_benchmarks = construct_benchmark_run_db(data)
-    weight_sum = 0.0
-    score_db = defaultdict(float)
-
-    for task, models in found_benchmarks.items():
-        for name, all_configs in models.items():
-            weight = b_weight[task] * (1.0/len(all_configs))
-            for mean, config, benchmark in all_configs:
-                weight_sum += weight
-                benchmark_score = weight * math.log(norm[benchmark] / mean)
-                score_db[config] += benchmark_score
-
-    assert abs(weight_sum - 1.0) < 1e-6, f"Bad configuration, weights don't sum to 1, but {weight_sum}"
-    return score_db
-
-def compute_score(weight_db, data, target, norm):
-    """
-    This API calculates the total score for all the benchmarks
-    that was run  by reading the data (.json) file.
-    The weights are then calibrated to the target score.
-
-    # TODO:
-    # score_db: score_per_config for now is just calculated.
-    # This needs to passed to plot_sweep.py to plot all the
-    # scores per config.
-    """
-    score = 0.0
-    score_db = get_score_per_config(data, weight_db, norm)
-
-    for config, scores in score_db.items():
-        score += scores
-        score_db[config] = target * math.exp(scores * 0.125)
-
-    score = target * math.exp(score)
-    return score
+class TorchBenchScore:
+    def __init__(self, spec=SPEC_FILE_DEFAULT, target=TARGET_SCORE_DEFAULT):
+        self.spec = spec
+        self.target = target
+        self.weights = None
+        self.norm = None
+
+    def setup_weights(self):
+        """
+        Calculates the static benchmark weights by iterating the spec
+        file and constructs a dictionary with (key, value) pair
+        is (task, weight_for_benchmark_per_task)
+
+        The spec here can be user defined .yaml file or if no spec is defined,
+        the default score.yaml is used.
+        """
+        # Load the spec file
+        with open(self.spec) as spec_file:
+            spec = yaml.full_load(spec_file)
+
+        self.weights = defaultdict(float)
+        category_spec = spec['hierarchy']['model']
+        domain_weight = 1.0/ len(category_spec)
+        for domain in category_spec:
+            tasks = category_spec[domain]
+            task_weight = 1.0 / len(tasks)
+            for task in tasks:
+                benchmarks = tasks[task]
+                benchmark_weight = 1.0 / len(benchmarks)
+                self.weights[task] = domain_weight * task_weight * benchmark_weight
+
+    def setup_benchmark_norms(self, data):
+        """
+        Helper function which gets the normalization values per benchmark
+        by going through the reference data file.
+
+        This reference data can be user provided or by default the first file
```

Comment:
try to decouple the comments here from the context of the plot code where it's currently being called. e.g. no point mentioning 'the first file' here - all this comment should care about is that you provide the reference data.  (and you could rename data to ref_data to make that clearer)

Reply:
also wondering if it makes sense for the user to call setup_benchmark_norms, or should it be called by __init__?  Is there a valid case to construct TorchBenchScore without providing ref_data? if so, this is good; if not, maybe just force ref_data be provided for __init__ and call this automatically

Reply:
I have added ref_data in `init`. I have created  `TORCHBENCH_V0_SCORE` which points to `torchbench_0.0.yaml`.I know we discussed creating it as object rather than a variable like this, but somebody who wants to use this as a reference data can easily import and pass this as the ref_data in the init of the Torchbench class. I have also created a placeholder for `TORCHBENCH_V1_SCORE`, so once available we just have to update the name of the file. If we create a TorchBenchScore Object for v0_score, the one thing I was concerned about was the spec and target gets set to the default spec and target values that was used while constructing the v0 score object. If someone wants to even try out a different spec or target values that wouldn't be possible. Wdyt?

Reply:
Well, I can provide a configure method inside TorchBenchScore class to manipulate the spec and target values if we want to go by this approach.

Reply:
I think it's more important to make it easy to grab the already configured v0 or v1 score, not to customize it.  The intent behind the v0 or v1 score is to make sure you're doing it _exactly_ the same way as CI.  I think it's a bad idea to let people think they are computing a 'customized v0 score' since any change makes it really not a v0 score anymore.

for the case of plotting relative score sweeps, it's more important to be flexible.  You can either use the ref_data as the source of truth (assume all the models present in it should be part of the score), or possibly do that but also add an exclude set so that you can make a relative score plot that skips some 'bad' models in your data.  This can be useful when coming up with a new score formula for example.

Reply:
Agreed and fixed except exclude_list.  For the exclude_list, I'll open up a follow up PR for this.